Types: make Interface a trait

gterzian opened the pull request:
In the light of https://github.com/massalabs/massa/pull/1995

gterzian commented:
cc @yvan-sraka

AureliaDolo commented on src/execution_impl.rs:
```diff
@@ -54,18 +52,18 @@ fn call(env: &Env, address: i32, function: i32) -> i32 {
 /// given intefrace, an operation number limit and a webassembly module
 ///
-fn create_instance(limit: u64, module: &[u8], interface: &Interface) -> Result<Instance> {
+fn create_instance(limit: u64, module: &[u8], interface: &Box<dyn Interface>) -> Result<Instance> {
```
Is clippy right ?

adrien-zinger replied:
Nop I don't think so.

gterzian replied:
From the CI it seems ok.

AureliaDolo commented on src/types.rs:
```diff
@@ -13,41 +13,56 @@ pub struct Response {
 
-#[derive(Clone)]
-pub struct Interface {
+pub trait InterfaceClone {
```
You define that trait so interface can kinda implement clone ?

gterzian replied:
Yes, so we can implement `Clone` for `Box< dyn Interface>`, so that it can be a member of all those structures that derive `Clone`.

adrien-zinger commented:
@gterzian I've rebased your code on this branch because we moved things, refacto, etc... and because this project is tiny and nervous I though you would like to continue to work with your code and a clean history! Feel free to use it or not, you can force push on your PR, and start earlier the integration it in massa. :wink: 

:clap: Look good to me btw, now I get it haha :pray: 

Rebased branch:
https://github.com/massalabs/sc-runtime/tree/use_trait_for_interface_rebased

gterzian commented:
> @gterzian I've rebased your code on this branch because we moved things, refacto, etc... and because this project is tiny and nervous I though you would like to continue to work with your code and a clean history! Feel free to use it or not, you can force push on your PR, and start earlier the integration it in massa. 😉
> 
> 👏 Look good to me btw, now I get it haha 🙏
> 
> Rebased branch: https://github.com/massalabs/sc-runtime/tree/use_trait_for_interface_rebased

@adrien-zinger thanks! 

Continuing the discussion at https://github.com/massalabs/sc-runtime/pull/35

adrien-zinger commented:
I close this PR for duplication!
